### PR TITLE
feat: added role & permissions resource to provider

### PR DESCRIPTION
**Pull request by williansouzagonc**
This PR includes 2 new resource to the EVA TF provider:

- Role 
- Role Permissions

### CODE REVIEW

File: internal/eva/role.go [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+
+type GetRoleRequest struct {
+	ID int64 `json:"ID"`
+}
+
+type Role struct {
+	Name                  string                    `json:"Name"`
+	UserType              int64                     `json:"UserType,omitempty"`
+	Code                  string                    `json:"Code,omitempty"`
+	ScopedFunctionalities []RoleScopedFunctionality `json:"ScopedFunctionalities"`
```
Comment (williansouzagonc) [inline internal/eva/role.go]: I created this struct `RoleScopedFunctionality` which is the same as `ScopedFunctionality` inside `role_permissions.go`, is a good practise to re-use it ? or do I have to always create a new struct ? @mad-it

File: internal/eva/role_permissions.go [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+	if resp.StatusCode() != 200 {
+		tflog.Info(ctx, "Request failed", "Status code", resp.StatusCode(), "body", resp.String())
+
+		return nil, errors.New("Request failed.")
+	}
+
+	tflog.Debug(ctx, "Request info", "Status code", resp.StatusCode(), "body", resp.String())
+
+	var jsonResp DetachFunctionalitiesFromRoleResponse
```
Comment (williansouzagonc) [inline internal/eva/role_permissions.go]: I think we don't really care to response here, so I think can be removed

File: internal/eva/role.go (hunk 1 above)
Comment (mongrelion) [reply internal/eva/role.go]: I would say that even if the fields are the same I would keep them separate for now because we don't know if the underlying API re-uses them or if they are different. If they are different we run the risk of them changing under the hood and that would break two of our dependencies

File: examples/resources/eva_role.tf [hunk 3]
```diff
@@ -0,0 +1,6 @@
+
+resource "eva_role" "example" {
+  name          = "My Example Role"
+  user_type     = 1
```
Comment (mad-it) [inline examples/resources/eva_role.tf]: image

If I remember correctly, UserType is a bit-wise operator. Is there a way we can harden this a bit more to not allow invalid values? 🤔

File: internal/eva/role_permissions.go (hunk 2 above)
Comment (mad-it) [reply internal/eva/role_permissions.go]: Have this in multiple places as well. I just added it as placeholder in case we want to do something with a response that we might get.

File: examples/resources/eva_role.tf (hunk 3 above)
Comment (williansouzagonc) [reply examples/resources/eva_role.tf]: How do we keep this list updated since we have no control under it ? I'd say for us to leave all validations in EVA API and we just proxy these values to them. what do you think ?

File: internal/provider/provider.go [hunk 4]
```diff
@@ -23,6 +23,10 @@ type provider struct {
 	// provider is built and ran locally, and "test" when running acceptance
 	// testing.
 	version string
+
+	apiEndpoint string
```
Comment (williansouzagonc) [inline internal/provider/provider.go]: Small tricky to be able to set up provider configuration for acceptance test. Terraform framework doesn't support default values for provider attributes yet, once they have it implemented we should be able to fetch these values from environment variables.

File: internal/provider/resource_eva_role_permissions.go [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+)
+
+type rolePermissionsType struct{}
+
+func (t rolePermissionsType) GetSchema(ctx context.Context) (tfsdk.Schema, diag.Diagnostics) {
+	return tfsdk.Schema{
+		MarkdownDescription: "Eva role permissions configuration.",
+
+		Attributes: map[string]tfsdk.Attribute{
+			"id": {
```
Comment (williansouzagonc) [inline internal/provider/resource_eva_role_permissions.go]: To be able to run acceptance tests every resource must have a `id` attribute (even when it's not being used), `role permissions` doesn't have a unique ID so for now I'm setting `RoleID` into id attribute.

https://www.terraform.io/plugin/framework/acctests#implement-id-attribute

File: examples/resources/eva_role.tf [hunk 6]
```diff
@@ -0,0 +1,6 @@
+
```
Comment (mad-it) [inline examples/resources/eva_role.tf]: newline not really needed.
```suggestion
```

File: examples/resources/eva_role_permissions.tf [hunk 7]
```diff
@@ -0,0 +1,11 @@
+
```
Comment (mad-it) [inline examples/resources/eva_role_permissions.tf]:
```suggestion
```

File: internal/eva/role.go (hunk 1 above)
Comment (mad-it) [reply internal/eva/role.go]: Within the `Role` scope, `ScopedFunctionalities` is not being managed. So this field can be completely omitted from the response since we are not really using it anywhere upstream.

**Review by mad-it — COMMENTED** [review]
PR looks good from the EVA client side, however have you considered the following TF schema:
```tf
resource "eva_role" "example" {
  name      = "My Example Role"
  user_type = 1
  code      = "my_example_role"
  scoped_functionalities = [
    {
      functionality      = "FunctionalityName"
      scope              = 1
      requires_elevation = false
    }
  ]
}
```

With this TF schema you call combine the underlying HTTP calls into a single resource which looking at the HTTP calls are fully intertwined with each other.
For example, you are introducing an "id" field which is essentially a hack to satisfy tests. This would then be unnecessary.

File: .github/workflows/test.yml [hunk 8]
```diff
@@ -6,10 +6,10 @@ name: Tests
 on:
```
Comment (mad-it) [inline .github/workflows/test.yml]: All the changes in this file should not be part of this PR. Please remove them and let this PR only focus on the role and permission resource.

File: internal/provider/provider.go [hunk 9]
```diff
@@ -42,14 +42,29 @@ func (p *provider) Configure(ctx context.Context, req tfsdk.ConfigureProviderReq
...
+		data.Token.Value = os.Getenv("EVA_API_TOKEN")
+	}
+
+	// required if still unset
+	if data.Endpoint.Value == "" {
+		resp.Diagnostics.AddError("No valid eva endpoint provided.", "A valid api endpoint is needed to authenticate on eva")
+		return
+	}
+
+	if data.Token.Value == "" && (data.Username.Null || data.Password.Null) {
```
Comment (mad-it) [inline internal/provider/provider.go]: Please remove these changes as well so the PR can focus only on the resource.

File: internal/provider/provider_test.go [hunk 10]
```diff
@@ -1,6 +1,7 @@
 package provider
```
Comment (mad-it) [inline internal/provider/provider_test.go]: same on this file.

File: internal/provider/resource_eva_role.go [hunk 11]
```diff
@@ -0,0 +1,270 @@
...
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-framework/tfsdk"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	"github.com/hashicorp/terraform-plugin-go/tftypes"
+	"github.com/hashicorp/terraform-plugin-log/tflog"
+
+	"github.com/mad-it/terraform-provider-eva/internal/eva"
+)
+
+type roleProviderType struct{}
```
Comment (mad-it) [inline internal/provider/resource_eva_role.go]: Maybe go with something like this since this is not a `provider`

```suggestion
type roleType struct{}
```

File: internal/provider/resource_eva_role.go [hunk 12]
```diff
@@ -0,0 +1,270 @@
...
+	data.ID = types.Int64{Value: createdRole.ID}
+
+	_, attachPermissionsToRoleErr := r.provider.evaClient.AttachFunctionalitiesToRole(ctx, eva.AttachFunctionalitiesToRoleRequest{
+		RoleID:                data.ID.Value,
+		ScopedFunctionalities: data.getListOfFunctionalities(),
+	})
+
+	if attachPermissionsToRoleErr != nil {
+		resp.Diagnostics.AddError("Creating role permissions failed.", fmt.Sprintf("Unable to create role permisions, got error: %s", attachPermissionsToRoleErr))
+		return
```
Comment (mad-it) [inline internal/provider/resource_eva_role.go:171]: Here you are returning, however the `role` was successfully created. That would lead to orphaned roles in case attachment fails. Not sure this is desired behaviour.

File: internal/provider/resource_eva_role.go [hunk 13]
```diff
@@ -0,0 +1,270 @@
...
+
+	_, attachErr := r.provider.evaClient.AttachFunctionalitiesToRole(ctx, eva.AttachFunctionalitiesToRoleRequest{
+		RoleID:                data.ID.Value,
+		ScopedFunctionalities: data.getListOfFunctionalities(),
+	})
+
+	if attachErr != nil {
+		resp.Diagnostics.AddError("Updating role permissions failed.", fmt.Sprintf("Unable to attach new role permissions, got error: %s", detachErr))
+		return
+	}
```
Comment (mad-it) [inline internal/provider/resource_eva_role.go:262]: hmmmm... Now I see what you meant earlier. Remove everything and re-add everything 🤔

File: internal/provider/resource_eva_role.go [hunk 14]
```diff
@@ -0,0 +1,270 @@
...
+						"scope": {
+							MarkdownDescription: "functionality scope",
+							Required:            true,
+							Type:                types.Int64Type,
+						},
+						"requires_elevation": {
+							MarkdownDescription: "whether functionality requires elevation or not",
+							Required:            true,
+							Type:                types.BoolType,
+						},
```
Comment (mad-it) [inline internal/provider/resource_eva_role.go]: image
Can we set default to false for this one? That way we match current importer functionality and dont need to specify it for each role.

Comment (williansouzagonc) [reply internal/provider/resource_eva_role.go]: Good catch, I can implement a modifier for the case value is not present in the plan configuration.

File: internal/provider/resource_eva_role.go (hunk 12 above)
Comment (williansouzagonc) [reply internal/provider/resource_eva_role.go:171]: Okay, maybe we need to revert the changes (this means deleting the newly created rule)

Comment (mad-it) [reply internal/provider/resource_eva_role.go:171]: There are things I think we could do:
1. Log an error saying "roles is not fully configured, please run the plan again" but **not** doing the `return`. That way the state is still successfully stored and the ID is not lost.
2. Deleting like you mentioned, however this also has a chance of failing as well 🙈

File: internal/provider/resource_eva_role.go [hunk 15]
```diff
@@ -60,11 +60,19 @@ func (t roleProviderType) GetSchema(ctx context.Context) (tfsdk.Schema, diag.Dia
...
+							PlanModifiers: []tfsdk.AttributePlanModifier{
+								boolDefaultModifier{
+									Default: false,
+								},
+							},
 						},
 					},
-					tfsdk.ListNestedAttributesOptions{},
+					tfsdk.ListNestedAttributesOptions{
+						MinItems: 1,
```
Comment (mad-it) [inline internal/provider/resource_eva_role.go:74]: Not sure if its the same but `MinItems` has been removed from: https://github.com/hashicorp/terraform-plugin-go/pull/143

But I might also be referencing the wrong object.

File: internal/provider/resource_eva_role.go [hunk 16]
```diff
@@ -60,11 +60,19 @@ func (t roleProviderType) GetSchema(ctx context.Context) (tfsdk.Schema, diag.Dia
 						},
 						"requires_elevation": {
 							MarkdownDescription: "whether functionality requires elevation or not",
-							Required:            true,
+							Optional:            true,
+							Computed:            true,
```
Comment (mad-it) [inline internal/provider/resource_eva_role.go:64]: How come its computed? Because a default is set?

File: internal/provider/modifiers.go [hunk 17]
```diff
@@ -0,0 +1,46 @@
+package provider
```
Comment (mad-it) [inline internal/provider/modifiers.go:1]: Lawl... the removed Default so that you have to write your own? 🤦🏼

File: internal/provider/resource_eva_role.go (hunk 16 above)
Comment (williansouzagonc) [reply internal/provider/resource_eva_role.go:64]: Yeah, once you have modifiers, you need to mark `computed` as `true`

File: internal/provider/modifiers.go (hunk 17 above)
Comment (williansouzagonc) [reply internal/provider/modifiers.go:1]: hahahaha.. exactly what I thought, but I also understand that it gives you more flexibility to implement your own logic.

File: internal/provider/resource_eva_role.go (hunk 15 above)
Comment (williansouzagonc) [reply internal/provider/resource_eva_role.go:74]: I think it's not talking about same struct

File: internal/provider/modifiers.go (hunk 17 above)
Comment (mad-it) [reply internal/provider/modifiers.go:1]: At least have a basic subset of often used ones like default -_-

**Review by mad-it — APPROVED** [review]
